rocketpool-daemon/common/validator/key-recovery-manager/dry-run-key-recovery-manager.go

```diff
@@ -0,0 +1,149 @@
+package key_recovery_manager
+
+import (
+	"fmt"
+	"github.com/rocket-pool/node-manager-core/beacon"
+	"github.com/rocket-pool/node-manager-core/utils"
+	"github.com/rocket-pool/smartnode/v2/rocketpool-daemon/common/validator"
+	"golang.org/x/exp/maps"
+	"strings"
+)
+
+type DryRunKeyRecoveryManager struct {
+	manager *validator.ValidatorManager
+}
+
+func NewDryRunKeyRecoveryManager(m *validator.ValidatorManager) *DryRunKeyRecoveryManager {
+	return &DryRunKeyRecoveryManager{
+		manager: m,
+	}
+}
+
+func (d *DryRunKeyRecoveryManager) RecoverMinipoolKeys() ([]beacon.ValidatorPubkey, map[beacon.ValidatorPubkey]error, error) {
+	status, err := d.manager.GetWalletStatus()
+	if err != nil {
+		return []beacon.ValidatorPubkey{}, map[beacon.ValidatorPubkey]error{}, err
+	}
+
+	rpNode, mpMgr, err := d.manager.InitializeBindings(status)
+	if err != nil {
+		return []beacon.ValidatorPubkey{}, map[beacon.ValidatorPubkey]error{}, err
+	}
+
+	publicKeys, err := d.manager.GetMinipools(rpNode, mpMgr)
+	if err != nil {
+		return []beacon.ValidatorPubkey{}, map[beacon.ValidatorPubkey]error{}, err
+	}
+
+	recoveredCustomPublicKeys, unrecoverableCustomPublicKeys, _ := d.checkForAndRecoverCustomKeys(publicKeys)
+	recoveredPublicKeys, unrecoverablePublicKeys := d.recoverConventionalKeys(publicKeys)
+
+	allRecoveredPublicKeys := []beacon.ValidatorPubkey{}
+	allRecoveredPublicKeys = append(allRecoveredPublicKeys, maps.Keys(recoveredCustomPublicKeys)...)
+	allRecoveredPublicKeys = append(allRecoveredPublicKeys, recoveredPublicKeys...)
+
+	for publicKey, err := range unrecoverablePublicKeys {
+		unrecoverableCustomPublicKeys[publicKey] = err
+	}
+
+	return allRecoveredPublicKeys, unrecoverablePublicKeys, nil
+}
+
+func (d *DryRunKeyRecoveryManager) checkForAndRecoverCustomKeys(publicKeys map[beacon.ValidatorPubkey]bool) (map[beacon.ValidatorPubkey]bool, map[beacon.ValidatorPubkey]error, error) {
+
+	recoveredKeys := make(map[beacon.ValidatorPubkey]bool)
+	recoveryFailures := make(map[beacon.ValidatorPubkey]error)
+	var passwords map[string]string
+
+	keyFiles, err := d.manager.LoadFiles()
+	if err != nil {
+		return recoveredKeys, recoveryFailures, err
+	}
+
+	if len(keyFiles) > 0 {
```

Comment:
check `if len(keyFiles) == 0` and return early. Outdent the entire conditional below.